map.getStyle() modification issue

gavinr-maps opened the pull request:
This change returns a clone of the object so that when calling `.getStyle()` the user cannot inadvertently modify the internally stored style. 

- Fixes #4481
- Unit tests: see fe7957bc16b09061b2428cc0e6d18e68d2975880
- Document any changes to public APIs: no changes



## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->


 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

HarelM commented on src/style/style.ts:
```diff
@@ -483,7 +483,7 @@ export class Style extends Evented {
     private _serializeByIds(ids?: Array<string>): Array<LayerSpecification> {
 
-        const serializedLayersDictionary = this._serializedAllLayers();
+        const serializedLayersDictionary = clone(this._serializedAllLayers());
```
I would consider doing the clone only later, inside the "for" to only clone the needed layers.

HarelM replied:
Also, since this is also used to send the data to the workers, which should not be cloned I would consider adding a flag to this method to clone only if this flag is "on".
One last comment, can we use structuredClone instead of lodash clone?

tomhicks replied:
Agree with the flag suggestion. This is potentially very performance sensitive - we use getStyle on mousemove and have recently _removed_ a lodash cloning of our MSS style (in favour of a shallow clone because of how we were using it) and it saved us ~7ms per mousemove on some maps.

Another alternative is deep freezing, which would prevent mutation and put the cloning requirement onto users.

gavinr-maps replied:
It looks like `structuredClone` was added in Node v17.0 - where this library is currently supporting https://github.com/maplibre/maplibre-gl-js/blob/5332a1c8524dd08ddb0a40590e32312d467b85bd/package.json#L190
... so I'm not allowed to use `structuredClone`, am I? I do not see it used elsewhere in the codebase.

HarelM replied:
I see, you are right, it is not used elsewhere, we can decide to replace all the `clone` with this later on, let's keep it outside the scope of this PR.

HarelM commented on src/ui/map_tests/map_style.test.ts:
```diff
@@ -359,6 +359,32 @@ describe('#getStyle', () => {
         });
     });
 
+    test('returns the previous style even if modified', done => {
```
use `await once('load')` instead of using `done` as we are trying to reduce the number of "done" in the tests.

gavinr-maps replied:
thanks! changed it to use async/await instead.

HarelM commented:
Thanks for taking the time to open this PR!
I've added a few minor comments.

gavinr-maps commented:
@HarelM thank you for the feedback and your time maintaining this project. I have addressed your feedback and is ready for your re-review when you have some time. Thanks.

HarelM reviewed (approved):
Looks good THANKS!

anneb commented:
Is this a breaking change?
A way to get the list of all layers is:
```
map.getStyle().layers
```
When handling layers, it is legal to modify the layer metadata. Modified layer metadata should be kept with the layer, not with a clone.

My workaround:
```
const layers = myMap.getStyle().layers;
const unclonedLayers = layers.map(layer => myMap.getLayer(layer.id));
```

From this issue, I can not infer why it was required to modify getStyle() to return a clone, not the original.
Layer metadata is a good way to store application information with a layer, so shouldn't it be updatable?

neodescis commented:
> Is this a breaking change? A way to get the list of all layers is:
> 
> ```
> map.getStyle().layers
> ```
> 
> When handling layers, it is legal to modify the layer metadata. Modified layer metadata should be kept with the layer, not with a clone.
> 
> My workaround:
> 
> ```
> const layers = myMap.getStyle().layers;
> const unclonedLayers = layers.map(layer => myMap.getLayer(layer.id));
> ```
> 
> From this issue, I can not infer why it was required to modify getStyle() to return a clone, not the original. Layer metadata is a good way to store application information with a layer, so shouldn't it be updatable?

Actually, getStyle().layers is a list of layer specifications, not the layer implementations themselves. The call to getLayer() in your example is retrieving the implementation, not the specification. The two are not interchangeable. 

Generally, updating an existing layer is done through getLayer(). Updating the original style or layer specification instances directly after they have been incorporated into the current map style isn't really going to work anyway, as nothing is watching for those changes.
